File: archive/Euler822/Euler822.cpp
#include <iostream>
#include <math.h>
#include <cmath>
#include <ctime>
#include <tuple>
#include <iomanip>
#include "CL_Arithmetics.h"

#define FOR(a, b, c) for(int a = b; a < c; a++)

long long int sum_b(int N, double K){//This gives the smallest number of steps so that all elements in the list are above e^2^K
    long long int ans = 0;
    FOR(a, 2, N+1) ans += ceil(K - log2(log(a)));
    return ans;
}


long double tiro(int N, long long int M){//This gives log2(log(min({2, ..., N}^M)) where {2, ..., N}^M is the set resulting after M operations
    long double uBound = N, lBound = N, mPoint, PREC = .000001;
    while(sum_b(N, uBound) < M ) uBound *= 2; 
    while(sum_b(N, lBound) > M) lBound /= 2;
    while(abs(uBound - lBound) > PREC){
        mPoint = (uBound + lBound ) / 2;
        std::cout << std::setprecision(std::numeric_limits<long double>::max_digits10) << "bounds: " << mPoint << " :: " << sum_b(N, mPoint) << " -- " << lBound << " :: " << sum_b(N, lBound) << " - " << uBound  << " :: " << sum_b(N, uBound) << std::endl;
        if(sum_b(N, mPoint) < M) lBound = mPoint;
        else uBound = mPoint;
    }
    return lBound;
}



int main(){
    clock_t time_req = clock();
    int N = 1;
    FOR(i, 0, 1) N *= 10;
    long long int M = 1;
    FOR(i, 0, 2) M *= 10;
    std::cout << M << ", " << N << std::endl;
    long double x = tiro(N, M);
    //std::cout << sum_b(N, 102.20312910620184) << " :: "  << sum_b(N, 102.20312910620294) << std::endl;
    long long int F = sum_b(N, x);
    std::cout << std::setprecision(std::numeric_limits<long double>::max_digits10) << x << " is reached after " << F << " steps, now we try to reach " << M << std::endl;
    long long int sum = 0, MOD = 1234567891;
    FOR(a, 2, N+1){
        int pow = PowerMod(a, PowerMod(2, ceil(x - log2(log(a))), MOD-1), MOD);
        sum += pow;
        sum %= MOD;
    }
    std::cout << "Current sum is " << sum << std::endl;

    //Find smallest element of the list after F steps
    long double min = x+1.0001;
    int argmin = 1;
    FOR(a, 2, N+1) if(min > log2(log(a)) + ceil(x - log2(log(a)))) {
        min = ceil(x - log2(log(a))) + log2(log(a));
        argmin = a;
    }
    std::cout << argmin << " is the smallest with power " << ceil(x - log2(log(argmin))) << std::endl;

    long long int smallest_power = PowerMod(argmin, PowerMod(2, ceil(x - log2(log(argmin))), MOD-1), MOD);
    std::cout << smallest_power << " is the power " << (smallest_power*smallest_power)%MOD << " the square." << std::endl;
    sum += ((M - F + MOD)%MOD)*(((smallest_power*smallest_power)%MOD - smallest_power + MOD)%MOD);
    sum %= MOD;

    std::cout << "Total sum = " << sum << std::endl;
    std::cout << ((double)(clock() - time_req))/1000 << " seconds" << std::endl;
    return 0;
}

File: archive/Euler822/Euler822.py
#Runs in 67s
#Uses very high precision
#Done on the 10/10/2023


import time
import math
from decimal import *
start = time.time()

getcontext().prec = 40

def sum_b(N, K):
    ans = 0
    for i in range(2, N+1):
        ans += math.ceil(K - ll(i))
    return ans

def tiro(N, M):
    uBound = Decimal(N)
    lBound = Decimal(N)
    PREC = Decimal(.1**6)
    while(sum_b(N, uBound) < M):
        uBound *= 2
    while(sum_b(N, lBound) > M):
        lBound /= 2
    while(uBound - lBound > PREC):
        mPoint = (uBound + lBound)/2
        aim = sum_b(N, mPoint)
        print(lBound, " / " , mPoint, " / " , uBound, " :::: ", aim)
        if(aim < M):
            lBound = mPoint
        else:
            uBound = mPoint
    return lBound


def PowerMod(a, n, p):
    #print("a = ", a, " | n = ", n, " | p = ", p)
    if n < 2:
        if n == 1:
            return a%p
        if n == 0:
            return 1
    ans = PowerMod((a*a)%p, n//2, p)
    if(n%2 == 1):
        ans *= a
        ans %= p
    return ans

def squareNTimes(a, k, mod):
    return pow(a, pow(2, k, mod-1), mod)

"""
def squareNTimes(a, pow, mod):
    ans = a
    for i in range(pow):
        ans *= ans
        ans %= mod
    return ans"""

def ll(n):
    return Decimal(n).ln().ln()/Decimal(2).ln()

N = 10**4
M = 10**16
x = tiro(N, M)
print("x = ", x)
F = sum_b(N, x)
sum = 0
MOD = 1234567891
for i in range(2, N+1):
    expon = squareNTimes(i, math.ceil(x - ll(i)), MOD)
    sum += expon

min = Decimal(x+1)
for i in range(2, N+1):
    n = math.ceil(x - ll(i)) 
    print(i, " was exponentiated ", n, " times.")
    if(min > n + ll(i)):
        min = math.ceil(x - ll(i)) + ll(i)
        argmin = i

print("Argmin = ", argmin, " with min = ", min)
smallest_power = squareNTimes(argmin, math.ceil(x - ll(argmin)), MOD)
sum += (M - F + MOD)*((smallest_power*smallest_power)%MOD - smallest_power + MOD)

print("smallest power = ", smallest_power, " with M - F = ", M - F)

sum %= MOD

print("Answer = ", sum)
end = time.time()
print("Time elapsed ", end - start, " seconds")